### PR TITLE
Notebook combine columns

### LINKED ISSUE #41178: Text column combination modal in the notebook editor
Implement the text column combination modal for the notebook editor.

See [Figma](https://www.figma.com/file/Rh7vrNTZX4TPAGC1fDtS55/QB-expressivity-and-column-actions%3A-extract-and-combine?type=design&mode=design&t=fkudaF5sAbXXQRif-0)

Screenshot 2024-04-12 at 11 42 07

### PR DESCRIPTION
**Pull request by romeovs**
Closes https://github.com/metabase/metabase/issues/41178

### Description

Adds the combine column shortcut to the expression editor popover. 

A lot of the work here is duplicated from https://github.com/metabase/metabase/pull/40082. There will be a follow up PR to merge and refactor the code.

What this PR does _not_ do (yet):
- Add the shortcut to the expression suggestions, this work is tracked elsewhere: https://github.com/metabase/metabase/pull/41578

### How to verify

Describe the steps to verify that the changes are working as expected.

1. New question → Sample Dataset → Orders
2. Add custom column
3. Click the `Combine` button
4. Play around with the combine popover to generate different expressions

### Demo

https://github.com/metabase/metabase/assets/1250185/12666698-af16-4274-bd34-92b0e5ff95aa

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/ui/components/inputs/TextInput/TextInput.styled.tsx [hunk 1]
```diff
@@ -1,22 +1,10 @@
 import type { MantineThemeOverride } from "@mantine/core";
```
Comment (romeovs) [inline frontend/src/metabase/ui/components/inputs/TextInput/TextInput.styled.tsx:1]: These changes have a wide blast radius. I'm happy to pull them out into a separate PR if that makes it easier to review.

The idea is that we make `Input.Wrapper` available so we can use it to build our own form components. For that to work the custom wrapper styles need to be applied to it too, and not just to the `wrapper` when used in `Input`.

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        opened={open}
+        onClose={handleClose}
+        onOpen={handleOpen}
+        closeOnEscape
+        closeOnClickOutside
+        width="target"
+        returnFocus
+      >
+        <Popover.Target>
+          <Button
```
Comment (romeovs) [inline frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.tsx:100]: We are emulating a select here with a button.

The initial idea (by Kamil) was to use Mantine's Select component with a custom `dropdownComponent` prop.

Mantine's `Select` does not allow the focus to go to the dropdown component (amongst other issues). It is very hacky and fragile to try and hack the custom dropdown into this component.

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.module.css [hunk 3]
```diff
@@ -0,0 +1,20 @@
+.button {
+  justify-content: space-between !important;
+  font-weight: normal;
+}
+
+button.root {
```
Comment (romeovs) [inline frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.module.css:6]: `button` is here for specificity, to avoid `!important` in the properties that are being overwritten.

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnAndSeparatorRow.module.css [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  }
+}
+
+.column {
+  flex: 1;
+}
+
+.remove {
+  flex: 0 0 auto;
+  border-color: transparent !important;
```
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnAndSeparatorRow.module.css:15]: we agreed not to use important, but either to introduce a new variant or fix theme

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.module.css (hunk 3 above)
Comment (uladzimirdev) [reply frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.module.css:6]: here is a [thread](https://metaboat.slack.com/archives/C0645JP1W81/p1712670698147929)

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+  function handleKeyDown(evt: KeyboardEvent<HTMLButtonElement>) {
+    if (evt.key === "Enter") {
+      setOpen(true);
+    }
+  }
+
+  const dropdown = (
+    <FocusTrap active={open}>
+      <div onBlur={handleBlur}>
```
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.tsx:66]: div doesn't have focus by default, even doesn't bubble, so what's the idea to have a handler here?

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.tsx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        returnFocus
+      >
+        <Popover.Target>
+          <Button
+            ref={button}
+            onMouseDownCapture={handleButtonClick}
+            onKeyDown={handleKeyDown}
+            fullWidth
+            classNames={{
+              root: classNames(styles.root, open && styles.open),
```
Comment (uladzimirdev) [inline frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.tsx:106]:
```suggestion
              root: classNames(styles.root, {[styles.open]: open}),
```

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.module.css (hunk 3 above)
Comment (romeovs) [reply frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnInput.module.css:6]: Do you want me to set up a custom variant for this? 

This variant is only going the be used in this instance, so I think might be overkill.

**Review by kamilmielnik — COMMENTED** [review]
I had a quick look at this PR - nice work! :+1:

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnAndSeparatorRow.module.css [hunk 7]
```diff
@@ -0,0 +1,22 @@
+.separator {
+  flex: 0 0 auto;
+
+  input {
+    font-family: Monaco, monospace;
```
[the remark this answers is not in the document]
Comment (romeovs) [reply frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnAndSeparatorRow.module.css:5]: I've discussed it with @AlessioLaiso and he agrees a monospace font would be better.

We don't have reusable classes for monospace fonts however, but we can add those of course!

File: frontend/src/metabase/ui/components/typography/Text/Text.styled.tsx [hunk 8]
```diff
@@ -5,6 +5,16 @@ export const getTextOverrides = (): MantineThemeOverride["components"] => ({
     defaultProps: {
       color: "text-dark",
       size: "md",
+      variant: "default",
+    },
+    variants: {
+      default: () => ({}),
```
Comment (romeovs) [inline frontend/src/metabase/ui/components/typography/Text/Text.styled.tsx:11]:
```suggestion
```

File: frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnAndSeparatorRow.tsx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    setHasFocus(false);
+  }
+
+  return (
+    <>
+      <TextInput
+        className={styles.separator}
+        label={t`Separator`}
+        value={value}
+        w={110}
```
Comment (nemanjaglumac) [inline frontend/src/metabase/query_builder/components/expressions/CombineColumns/ColumnAndSeparatorRow/ColumnAndSeparatorRow.tsx:121]:
```suggestion
        w={rem(110)}
```

**Review by nemanjaglumac — APPROVED** [review]
It's hard to follow all changes in the code in this PR, so I was focusing mostly on the functionality. I checked out the branch and tested it locally - everything seems to be working just fine and without any glitches.

I couldn't find a way to break it.
